### PR TITLE
vim: Add Smart Relative Line Number

### LINKED ISSUE #16514: Vim: absolute numbering in any mode except `insert` mode
### Check for existing issues

- [X] Completed

### Describe the feature

I am using the following vim configuration that allows relative numbering in any mode except in insert mode:

```vim9
set relativenumber # Relative line numbering

autocmd InsertEnter * :set norelativenumber
autocmd InsertLeave * :set relativenumber
```

For now, Zed forces me to choose between relative line numbering or absolute line numbering.

I am not sure how this feature should be introduced. I see two possible approaches:

- add a new option `line_numbers` that takes `absolute`, `relative` and `both` as values. `both` could be the behavior I described here. Unfortunately I didn't find a better name...
- add a way to change settings using contexts (the same concept used by keymaps).
  We could imagine something like:
  ```json
  {
    "context": "vim_mode == insert",
    "settings": { "relative_line_numbers": false }
  },
  {
    "context": "vim_mode != insert",
    "settings": { "relative_line_numbers": true }
  }
  ```
  However, this could introduce many issues. For example: should we accept any settings like theme, ...?

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by 0x2CA**
Closes #16514

Release Notes:

- Added Vim: absolute numbering in any mode except `insert` mode

### CODE REVIEW

**Comment by ConradIrwin** [comment]
@0x2CA Thank you for this!

A few changes I think we should make:
* We should call `editor.set_relative_line_numbers()` inside `Vim::sync_vim_settings` instead of in switch_mode. This will handle edge-cases like enabling/disabling vim mode.
* We should maintain the state per-editor not globally, so we need an instance property on the `Editor` struct `use_relative_line_numbers: Option<bool>`. When it's set we ignore the user's setting, and so in `Vim::deactivate` we should set it back to `None`.
* If we want to copy https://github.com/jeffkreeftmeijer/vim-numbertoggle more closely, we should also disable relative line numbers in `Vim::blurred` (but up to you, I don't mind either way).

**Comment by 0x2CA** [comment]
> @0x2CA Thank you for this!
> 
> A few changes I think we should make:
> 
> * We should call `editor.set_relative_line_numbers()` inside `Vim::sync_vim_settings` instead of in switch_mode. This will handle edge-cases like enabling/disabling vim mode.
> * We should maintain the state per-editor not globally, so we need an instance property on the `Editor` struct `use_relative_line_numbers: Option<bool>`. When it's set we ignore the user's setting, and so in `Vim::deactivate` we should set it back to `None`.
> * If we want to copy https://github.com/jeffkreeftmeijer/vim-numbertoggle more closely, we should also disable relative line numbers in `Vim::blurred` (but up to you, I don't mind either way).

hi, this is my first time writing Rust, I know nothing about Rust, I'm just winging it with my experience, please forgive my weird code, if you have better suggestions, please let me know, I'll study them, might need a little time

**Comment by 0x2CA** [comment]
@ConradIrwin 
review

**Comment by ConradIrwin** [comment]
@0x2CA Thanks for taking this on! I'm excited to have you writing code as well as reviewing features :D.

The Toggle command does not work well with the smart feature :/. I see why you had it just in switch mode, but that causes problems if you enable/disable vim mode, even if you just update it in `focused` and in `switch_mode`.

Given that you can show absolute numbers by switching to insert mode (or by opening the command palette) do we need a manual toggle too?

File: crates/vim/src/vim.rs [hunk 1]
```diff
@@ -1014,9 +1012,11 @@ impl Vim {
             editor.set_input_enabled(vim.editor_input_enabled());
             editor.set_autoindent(vim.should_autoindent());
             editor.selections.line_mode = matches!(vim.mode, Mode::VisualLine);
-            if VimSettings::get_global(cx).smart_relative_line {
+            if VimSettings::get_global(cx).toggle_relative_line_numbers {
                 let is_relative = vim.mode != Mode::Insert;
-                editor.set_relativg_line_numbers(Some(is_relative), cx)
+                editor.set_relative_line_number(Some(is_relative), cx)
+            } else {
+                editor.set_relative_line_number(None, cx)
```
Comment (0x2CA) [inline crates/vim/src/vim.rs:1019]: If it's not enabled, using the default settings will cause temporary switching to fail, so we should preserve the original logic, which is to prioritize reading the configuration when it's not enabled, and then user manual toggle.

**Comment by 0x2CA** [comment]
> Given that you can show absolute numbers by switching to insert mode (or by opening the command palette) do we need a manual toggle too?

I think it's still necessary, you can't casually stroll around in the file.

> If we want to copy https://github.com/jeffkreeftmeijer/vim-numbertoggle more closely, we should also disable relative line numbers in Vim::blurred (but up to you, I don't mind either way).

Because of the toggle issue, this function cannot be implemented, but I think it's better than not being able to use toggle at all.

If you want to achieve this function, you can't simply implement it by entering and leaving events. You need to know the previous vim that was entered, and only switch to another vim to update, so that the open command or panel won't be updated.

I think this is a bit difficult for me :(

**Comment by ConradIrwin** [comment]
Who knew it would be so tricky :D.

If we descope the blur/focus stuff then does this work?
* On switch_mode, call `set_use_relative_line_numbers`
* On settings change, call `set_use_relative_line_numbers` if either the vim_mode setting changes or the toggle_relative_line_numbers setting has changed.

We could add the tracking of which vim is active (the mode indicator already does this), but it sounds like you don't need that for v1.

**Comment by 0x2CA** [comment]
@ConradIrwin  

Of course, not implementing this function is actually the same as VSCodeVim's behavior, and there's no problem with that, it's just that adding this function would be better.

I implemented the blurred function based on the principle of the mode indicator.

review?

**Comment by ConradIrwin** [comment]
@0x2CA This is looking good, thank you! 

I pushed two small changes, the first is just stylistic, and the second ensures that when users edit their settings file we stay in sync.

Let me know if these look good to you, and I'll merge.

File: crates/vim/src/vim.rs [hunk 2]
```diff
@@ -635,30 +635,23 @@ impl Vim {
...
-            if VimSettings::get_global(cx).toggle_relative_line_numbers {
-                self.update_editor(cx, |_, editor, cx| {
-                    editor.set_relative_line_number(Some(true), cx)
-                });
             }
-            Vim::globals(cx).vim = Some(cx.view().downgrade());
+
+            self.update_editor(cx, |vim, editor, cx| {
+                let is_relative = vim.mode != Mode::Insert;
+                editor.set_relative_line_number(Some(is_relative), cx)
```
Comment (0x2CA) [inline crates/vim/src/vim.rs]: If written here, it will cause the command to return and forcibly reset, resulting in the toggle command failure.

**Comment by 0x2CA** [comment]
@ConradIrwin 
It looks great, more concise than mine, we can merge it.

**Comment by ConradIrwin** [comment]
Great, thank you!

**Comment by lougreenwood** [comment]
@0x2CA How is this used, I found this issue from the release notes, but I don't see the any example for how to enable it in the PR?

I see `toggle_relative_line_numbers` is a new setting - I didn't try it yet, but if this is the setting we need to enable it's not very descriptive and seems to be named more like an action than a setting.

Don't mean to seem nitpicky, excited to try this feature! :D

**Comment by 0x2CA** [comment]
> @0x2CA How is this used, I found this issue from the release notes, but I don't see the any example for how to enable it in the PR?
> 
> I see `toggle_relative_line_numbers` is a new setting - I didn't try it yet (still doing a `brew update` to get the latest preview), but if this is the setting we need to enable it's not very descriptive and seems to be named more like an action than a setting.
> 
> Don't mean to seem nitpicky, excited to try this feature! :D

I can understand your confusion. Initially, I used the setting name `smart relative line`, and after implementing the corresponding function, @ConradIrwin  optimized the corresponding code and renamed it, which is why the setting is named `toggle_relative_line_numbers `.

**Comment by lougreenwood** [comment]
Thanks for the explanation - naming is hard!

I wonder if since this seems to be a modal editor specific setting it would be more descriptive if it were called something like: `editor_mode_smart_relative_line_numbers`... 🤷‍♂️ 😄

**Pull request merged**